Docs: DPC++ Compiler (#17421)
LLNL/spack: lib/spack/docs/getting_started.rst

           spec: intel@15.0.24.4.9.3

^^^^^^^^^^
SYCL/DPC++
^^^^^^^^^^

The SYCL/DPC++ compiler is Intel and Codeplay's `staging area compiler <https://github.com/intel/llvm>`_ for upcoming Clang/LLVM SYCL support.

> This compiler is typically distributed with `oneAPI <https://software.intel.com/content/www/us/en/develop/articles/oneapi-repo-instructions.html>`_ and gets activated with ``source /opt/intel/inteloneapi/setvars.sh`` on Linux.

Reviewer: ```suggestion
This compiler is available in binary form via apt & yum packages, docker containers, or an intel installer with `oneAPI <https://software.intel.com/content/www/us/en/develop/tools/oneapi/download.html#basekit>`_ and can be activated with ``source /opt/intel/inteloneapi/setvars.sh`` on Linux.
```

Author: Thanks a lot for reviewing this, @rscohn2!

> This compiler is available in binary form via apt & yum packages, docker containers, or an intel installer with `oneAPI <https://software.intel.com/content/www/us/en/develop/tools/oneapi/download.html#basekit>`_ and can be activated with ``source /opt/intel/inteloneapi/setvars.sh`` on Linux.

Author: ```suggestion
This compiler is available in binary form via apt & yum packages, docker containers, or an intel installer with `oneAPI <https://software.intel.com/content/www/us/en/develop/tools/oneapi/download.html#basekit>`_ and can be activated with ``source /opt/intel/oneapi/setvars.sh`` on Linux.
```

One can record these activation steps by comparing set environment variables:

.. code-block:: console

   env > env.txt
   set +e

>    source /opt/intel/inteloneapi/setvars.sh

Author: ```suggestion
   source /opt/intel/oneapi/setvars.sh
```

   env > env_extras.txt
   diff env.txt env_extras.txt

All newly set environment hints will then be shown in this diff.
A typical entry in ``compilers.yaml`` looks like this:

.. code-block:: yaml

>    # DPC++ 2021.1-beta07 (2020.5.0.0604)

Author: ```suggestion
   # DPC++ 2021.1-beta09 (2020.8.0.0827)
```

   - compiler:
       spec: clang@11.0.0
       paths:

>          cc: /opt/intel/inteloneapi/compiler/latest/linux/bin/clang
>          cxx: /opt/intel/inteloneapi/compiler/latest/linux/bin/dpcpp

Author: ```suggestion
         cc: /opt/intel/oneapi/compiler/2021.1-beta09/linux/bin/clang
         cxx: /opt/intel/oneapi/compiler/2021.1-beta09/linux/bin/dpcpp
```

         f77:
         fc:
       flags: {}
       operating_system: ubuntu18.04
       target: x86_64
       modules: []
       environment:
         set:
           OCL_ICD_FILENAMES: "libintelocl_emu.so:libalteracl.so:/opt/intel/inteloneapi/compiler/latest/linux/lib/x64/libintelocl.so"
           ACL_BOARD_VENDOR_PATH: "/opt/Intel/OpenCLFPGA/oneAPI/Boards"
           INTELFPGAOCLSDKROOT: "/opt/intel/inteloneapi/compiler/2021.1-beta07/linux/lib/oclfpga"
           DPCPP_ROOT: "/opt/intel/inteloneapi"
           ONEAPI_ROOT: "/opt/intel/inteloneapi"
         prepend_path:

>            LD_LIBRARY_PATH: "/opt/intel/inteloneapi/compiler/latest/linux/lib:/opt/intel/inteloneapi/compiler/latest/linux/lib/x64:/opt/intel/inteloneapi/compiler/2021.1-beta07/linux/lib/oclfpga/host/linux64/lib:/opt/intel/inteloneapi/compiler/2021.1-beta07/linux/lib/oclfpga/linux64/lib:/opt/intel/inteloneapi/compiler/latest/linux/compiler/lib/intel64_lin:/opt/intel/inteloneapi/compiler/latest/linux/compiler/lib"

Author: Urgh, this is now three releases old and lives all in `/opt/intel/oneapi/` now

>            OCL_ICD_FILENAMES: "libintelocl_emu.so:libalteracl.so:/opt/intel/inteloneapi/compiler/latest/linux/lib/x64/libintelocl.so"
>            ACL_BOARD_VENDOR_PATH: "/opt/Intel/OpenCLFPGA/oneAPI/Boards"
>            INTELFPGAOCLSDKROOT: "/opt/intel/inteloneapi/compiler/2021.1-beta07/linux/lib/oclfpga"
>            DPCPP_ROOT: "/opt/intel/inteloneapi"
>            ONEAPI_ROOT: "/opt/intel/inteloneapi"
>          prepend_path:
>            LD_LIBRARY_PATH: "/opt/intel/inteloneapi/compiler/latest/linux/lib:/opt/intel/inteloneapi/compiler/latest/linux/lib/x64:/opt/intel/inteloneapi/compiler/2021.1-beta07/linux/lib/oclfpga/host/linux64/lib:/opt/intel/inteloneapi/compiler/2021.1-beta07/linux/lib/oclfpga/linux64/lib:/opt/intel/inteloneapi/compiler/latest/linux/compiler/lib/intel64_lin:/opt/intel/inteloneapi/compiler/latest/linux/compiler/lib"
>            CPATH: "/opt/intel/inteloneapi/dev-utilities/2021.1-beta07/include:/opt/intel/inteloneapi/compiler/latest/linux/include"
>            LIBRARY_PATH: "/opt/intel/inteloneapi/compiler/latest/linux/lib"
>            PATH: "/opt/intel/inteloneapi/dev-utilities/2021.1-beta07/bin:/opt/intel/inteloneapi/compiler/2021.1-beta07/linux/lib/oclfpga/bin:/opt/intel/inteloneapi/compiler/latest/linux/bin/intel64:/opt/intel/inteloneapi/compiler/latest/linux/bin:/opt/intel/inteloneapi/compiler/latest/linux/ioc/bin"

Author: ```suggestion
           OCL_ICD_FILENAMES: "libintelocl_emu.so:libalteracl.so:/opt/intel/oneapi/compiler/2021.1-beta09/linux/lib/x64/libintelocl.so"
           ACL_BOARD_VENDOR_PATH: "/opt/Intel/OpenCLFPGA/oneAPI/Boards"
           INTELFPGAOCLSDKROOT: "/opt/intel/oneapi/compiler/2021.1-beta09/linux/lib/oclfpga"
           ONEAPI_ROOT: "/opt/intel/oneapi"
         prepend_path:
           LD_LIBRARY_PATH: "/opt/intel/oneapi/compiler/2021.1-beta09/linux/lib:/opt/intel/oneapi/compiler/2021.1-beta09/linux/lib/x64:/opt/intel/oneapi/compiler/2021.1-beta09/linux/lib/emu:/opt/intel/oneapi/compiler/2021.1-beta09/linux/lib/oclfpga/host/linux64/lib:/opt/intel/oneapi/compiler/2021.1-beta09/linux/lib/oclfpga/linux64/lib:/opt/intel/oneapi/compiler/2021.1-beta09/linux/compiler/lib/intel64_lin:/opt/intel/oneapi/compiler/2021.1-beta09/linux/compiler/lib:/opt/intel/oneapi/tbb/2021.1-beta09/lib/intel64/gcc4.8"
           CPATH: "/opt/intel/oneapi/compiler/2021.1-beta09/linux/include:/opt/intel/oneapi/dev-utilities/2021.1-beta09/include:/opt/intel/oneapi/tbb/2021.1-beta09/include"
           LIBRARY_PATH: "/opt/intel/oneapi/compiler/2021.1-beta09/linux/lib:/opt/intel/oneapi/tbb/2021.1-beta09/lib/intel64/gcc4.8"
           PATH: "/opt/intel/oneapi/compiler/2021.1-beta09/linux/lib/oclfpga/llvm/aocl-bin:/opt/intel/oneapi/compiler/2021.1-beta09/linux/lib/oclfpga/bin:/opt/intel/oneapi/compiler/2021.1-beta09/linux/bin/intel64:/opt/intel/oneapi/compiler/2021.1-beta09/linux/bin:/opt/intel/oneapi/compiler/2021.1-beta09/linux/ioc/bin:/opt/intel/oneapi/dev-utilities/2021.1-beta09/bin"
```

           spec: intel@15.0.24.4.9.3

^^^^^^^^^^
SYCL/DPC++

> ^^^^^^^^^^

Reviewer 2: I don't think this belongs in the "getting started" section.

Maybe it could go in the longer section on intel tools?

Reviewer 2: https://spack.readthedocs.io/en/latest/build_systems/intelpackage.html?highlight=intel#intel-packages-in-spack

Author: Wasn't sure, that section reads:
```
Vendor-Specific Compiler Configuration

With Spack, things usually “just work” with GCC. Not so for other compilers. This section provides details on how to get specific compilers working.
```

It sets up ICC, PGI and NAG compilers in detail and lacks so far details on how to set environment vars for them. I think its activation details are unique enough to add value for the section.

Reviewer 3: I struggled with DPCPP just last week so this is well timed.

Reviewer: @frankwillmore Can you look at the compilers.yaml?

Reviewer 4: @rscohn2 That stanza looks pretty involved for an example. I've never had to add that much stuff, by ymmv. 
I stopped naming the compiler as 'clang' a while back, in favor of 'oneapi', because I think it's important to distinguish the intel version, as with the cray version of it. All of the environment stuff may well be necessary and valid for that particular installation, and I don't see anything fundamentally wrong, but imho it looks a bit cluttered. 

In my work, I've used icpx instead of dpcpp. afaik it will compile dpcpp code, just needs a few extra include paths. And I would set up the environment in a module, but this may just be a matter of taste.

Reviewer: I also stopped using dpcpp driver and instead add -fsycl to compile and link. All the cmake docs discourage changing the driver.
Unfortunately, you may also need -fscyl-unnamed-lambda for compile, but I hope they make that unnecessary.

The SYCL/DPC++ compiler is Intel and Codeplay's `staging area compiler <https://github.com/intel/llvm>`_ for upcoming Clang/LLVM SYCL support.
They are usually distributed with `oneAPI <https://software.intel.com/content/www/us/en/develop/articles/oneapi-repo-instructions.html>`_ and compilers are activated with ``source /opt/intel/inteloneapi/setvars.sh`` on Linux.

One can record these activation steps by comparing set environment variables:

.. code-block:: console

   env > env.txt
   set +e
   source /opt/intel/inteloneapi/setvars.sh
   env > env_extras.txt
   diff env.txt env_extras.txt

All newly set environment hints will then be shown in this diff.
A typical entry in ``compilers.yaml`` looks like this:

.. code-block:: yaml

   # DPC++ 2021.1-beta07 (2020.5.0.0604)
   - compiler:
       spec: clang@11.0.0
       paths:
         cc: /opt/intel/inteloneapi/compiler/latest/linux/bin/clang
         cxx: /opt/intel/inteloneapi/compiler/latest/linux/bin/dpcpp
         f77:
         fc:
       flags: {}
       operating_system: ubuntu18.04
       target: x86_64
       modules: []
       environment:
         set:

>            OCL_ICD_FILENAMES: libintelocl_emu.so:libalteracl.so:/opt/intel/inteloneapi/compiler/latest/linux/lib/x64/libintelocl.so

Author: @becker33 do I need to be careful with the `:`s here? maybe quote all strings or so?

Reviewer 2: I think you need to quote the strings with colons in them.

           ACL_BOARD_VENDOR_PATH: /opt/Intel/OpenCLFPGA/oneAPI/Boards
           FI_PROVIDER_PATH: /opt/intel/inteloneapi/mpi/2021.1-beta07//libfabric/lib/prov
           CLASSPATH: /opt/intel/inteloneapi/mpi/2021.1-beta07//lib/mpi.jar
           INTELFPGAOCLSDKROOT: /opt/intel/inteloneapi/compiler/2021.1-beta07/linux/lib/oclfpga
           NLSPATH: /opt/intel/inteloneapi/mkl/2021.1-beta07/lib/intel64/locale/%l_%t/%N
           DPCPP_ROOT: /opt/intel/inteloneapi
           ONEAPI_ROOT: /opt/intel/inteloneapi
           MKLROOT: /opt/intel/inteloneapi/mkl/2021.1-beta07
           TBBROOT: /opt/intel/inteloneapi/tbb/2021.1-beta07/env/..

>            I_MPI_ROOT: /opt/intel/inteloneapi/mpi/2021.1-beta07

Reviewer 2: Some of these environment variables don't properly belong as part of the compiler. The compiler should NOT know about I_MPI or MKL unless the user also requested one of those as an external package.

Author: Yeah, confused me as well... guess that's because I installed more basic packages. Will remove.

         prepend_path:
           LD_LIBRARY_PATH: /opt/intel/inteloneapi/tbb/2021.1-beta07/env/../lib/intel64/gcc4.8:/opt/intel/inteloneapi/mpi/2021.1-beta07//libfabric/lib:/opt/intel/inteloneapi/mpi/2021.1-beta07//lib/release:/opt/intel/inteloneapi/mpi/2021.1-beta07//lib:/opt/intel/inteloneapi/mkl/2021.1-beta07/lib/intel64:/opt/intel/inteloneapi/compiler/latest/linux/lib:/opt/intel/inteloneapi/compiler/latest/linux/lib/x64:/opt/intel/inteloneapi/compiler/2021.1-beta07/linux/lib/oclfpga/host/linux64/lib:/opt/intel/inteloneapi/compiler/2021.1-beta07/linux/lib/oclfpga/linux64/lib:/opt/intel/inteloneapi/compiler/latest/linux/compiler/lib/intel64_lin:/opt/intel/inteloneapi/compiler/latest/linux/compiler/lib

>            CPATH: /opt/intel/inteloneapi/tbb/2021.1-beta07/env/../include:/opt/intel/inteloneapi/mpi/2021.1-beta07//include:/opt/intel/inteloneapi/mkl/2021.1-beta07/include:/opt/intel/inteloneapi/dev-utilities/2021.1-beta07/include:/opt/intel/inteloneapi/compiler/latest/linux/include

Reviewer 2: Again, this is including MPI information that definitely should not be part of the compiler.
